### ISSUE TITLE
Loader leaks critical sections

### ISSUE THREAD

**Issue by scyganINTEL** [KhronosGroup/Vulkan-LoaderAndValidationLayers#2341]
When running any Vulkan app under ['Application Verifier'](https://docs.microsoft.com/en-us/windows-hardware/drivers/debugger/application-verifier) - with all basic checks selected, I am getting following error on `FreeLibrary(<vulkan-1.dll handle): 

```
=======================================
VERIFIER STOP 0000000000000201: pid 0x13CC: Unloading DLL containing an active critical section. 

	00007FFBAF88F7A0 : Critical section address. Run !cs -s <address> to get more information.
	00000151877DC650 : Critical section initialization stack trace. Run dps <address> to dump the stack trace.
	000001524081BFE0 : DLL name address.
	00007FFBAF790000 : DLL base address.
```

I have matched the address of the critical section to `loader_json_lock` mutex from https://github.com/KhronosGroup/Vulkan-LoaderAndValidationLayers/blob/874fbdaad9733e71274906e65475d716ffb07c70/loader/loader.c#L1909: 
```
void loader_initialize(void) {
    // initialize mutexs
    loader_platform_thread_create_mutex(&loader_lock);
    loader_platform_thread_create_mutex(&loader_json_lock);

    // initialize logging
```

This looks like a bug - both mutex-es from above are not destroyed anywhere. 

However, in the absence of DllMain (?)  it is unclear to me where to put the cleanup code - hence this issue.

**Comment by scyganINTEL** [comment]
I am seeing this on sdk-1.0.65 and latest master (6ce028589b360d98c1789762895aaf6462792ea4)

**Comment by lenny-lunarg** [comment]
When I try running cube from master, I actually get a different error from the Application Verifier:

```
VERIFIER STOP 0000000000000900: pid 0x6230: A heap allocation was leaked. 

	000001B395F42550 : Address of the leaked allocation. Run !heap -p -a <address> to get additional information about the allocation.
	000001B38A482C80 : Address to the allocation stack trace. Run dps <address> to view the allocation stack.
	000001B394F76FD8 : Address of the owner dll name. Run du <address> to read the dll name.
	00007FFFA6FC0000 : Base of the owner dll. Run .reload <dll_name> = <address> to reload the owner dll. Use 'lm' to get more information about the loaded and unloaded modules.
```

The call stack shows that this is happening when calling `FreeLibrary()` on the Intel driver. Checking the name of the dll that's creating the error gives:

```
>du 0x000001B394F76FD8
0x000001B394F76FD8  OPENGL32.dll.탐탐탐
```

I'm using a relatively old driver (it's whatever shipped from Windows Update). The device manager gives the date as 4/21/2017 and the version as 21.20.16.4664. Is there any specific driver I should try to run this on?

According the prompt I'm supposed to be able to continue the program by entering `go`, but that isn't working (I haven't used the Application Verifier in the past). Since I can't continue the program, I haven't been able to reproduce the error you reported.

That being said, I can tell just from reading the loader code that you're correct and we are leaking the two mutexes. I've created a branch that adds a DllMain on windows and uses the constructor and destructor attributes on linux so that the mutexes should get freed. The change is on branch `lenny_gh2341`, but I'm not going to merge that change until one of us is able to test that the problem is actually resolved on the branch. If you could either help me get past this error or test the branch yourself, then I can merge it in.

**Comment by jeffbolznv** [comment]
@lenny-lunarg, as a general rule, DLL_PROCESS_DETACH should return without freeing anything when the "reserved" parameter is non-NULL (process termination), and should only free things when it is NULL (FreeLibrary). MSDN has this to say about it:

"When handling DLL_PROCESS_DETACH, a DLL should free resources such as heap memory only if the DLL is being unloaded dynamically (the lpReserved parameter is NULL). If the process is terminating (the lpvReserved parameter is non-NULL), all threads in the process except the current thread either have exited already or have been explicitly terminated by a call to the ExitProcess function, which might leave some process resources such as heaps in an inconsistent state. In this case, it is not safe for the DLL to clean up the resources. Instead, the DLL should allow the operating system to reclaim the memory."

(https://msdn.microsoft.com/en-us/library/windows/desktop/ms682583(v=vs.85).aspx)

**Comment by scyganINTEL** [comment]
I have seen this in CTS, which loads (and frees) vulkan-1.dll dynamically. 

Yep, I am going to give `lenny_gh2341` branch a try. 

Agree with @jeffbolznv, that this should apply only to dynamic unload.

**Comment by scyganINTEL** [comment]
`lenny_gh2341` branch seems to have a typo: deletes `loader_lock` twice, does not delete  `loader_lock_json` here: https://github.com/KhronosGroup/Vulkan-LoaderAndValidationLayers/commit/e728287ebeec6071b4b4a1f7cbaa2a3d63e9d542#diff-5b4e070f5ce5cbb7bd1dd11f689d50a8R1927

Other than that (and the comment from @jeffbolznv that should be addressed), this works okay for me. 

I can see also other (non-mutex) related app verifier issues, but this deserves another issue / pr I am going to post soon.

**Comment by lenny-lunarg** [comment]
> as a general rule, DLL_PROCESS_DETACH should return without freeing anything when the "reserved" parameter is non-NULL (process termination), and should only free things when it is NULL

Thanks for pointing that out, you're definitely correct.

> `lenny_gh2341` branch seems to have a typo: deletes loader_lock twice, does not delete `loader_lock_json`

Yep, that's a typo.

I've now updated the `lenny_gh2341` branch to address both of these issues. I'll open a PR shortly to give anyone who wants to look over the changes a last chance before I merge.

**Comment by lenny-lunarg** [comment]
This bug is resolved by #2343.
